### PR TITLE
fix(sort): Treat comparison against undefined as undefined.

### LINKED ISSUE #164: [Bug]: Non-numeric values are always sorted as being larger than all numeric values
### My Home Assistant Version

Whether or not non-numeric values should be consider more or less than numeric values should be configurable.  Currently, numeric values are always considered to be greater than all numeric values.

### My lovelace configuration method

GUI

### What I am doing

Listing how full all the batteries are, sorted from least to most so that the batteries most in need of charge are at the top of the list.  I also want all batteries with unknown status to be at the top, so that the batteries that are of the most concern are up top.

### What I expect to happen

I expect to have unknown come before all numeric values.  Or I expect to be able to choose.

### What happened instead

unknown is always last.

### Minimal steps to reproduce

1. Load the yaml below when you have battery notes.
2. Intentionally disable one of the batteries.
3. Observe that the unknown battery is always grouped after the 100% batteries, or if reversed, before the 100%.
...

### Include any yaml code here

```YAML
type: custom:auto-entities
card:
  type: grid
  columns: 2
  square: false
card_param: cards
filter:
  template: >-
    {% set exclude_sensors = {} %}

    {% set label_entities_list =
      label_entities(exclude_sensors.label_id)
      if exclude_sensors.label_id is defined
         and label_entities(exclude_sensors.label_id) is not none
      else [] %}
    {% set entity_list = exclude_sensors.entity_id if exclude_sensors.entity_id
    is defined else [] %} {% if entity_list is string %}
      {% set entity_list = [entity_list] %}
    {% endif %} {% set all_exclude_sensors = label_entities_list + entity_list
    %}

    {% set ns = namespace(cards=[], seen=[]) %}

    {# 1. Battery Plus "low" is on — show the numeric battery sensor when it
    exists #} {% for state in states.binary_sensor
         | rejectattr('attributes.device_class', 'undefined')
         | selectattr('attributes.device_class', '==', 'battery')
         if state.entity_id.endswith('_battery_plus_low')
            and state.state in ['on', 'low']
            and state.entity_id not in all_exclude_sensors %}
      {% set sensor_id = 'sensor.' ~ state.object_id.removesuffix('_low') %}
      {% set entity = sensor_id if states[sensor_id] is defined else state.entity_id %}
      {% if entity not in all_exclude_sensors and entity not in ns.seen %}
        {% set ns.seen = ns.seen + [entity] %}
        {% set ns.cards = ns.cards + [{
          'type': 'custom:bubble-card',
          'card_type': 'button',
          'button_type': 'state',
          'entity': entity,
          'grid_options': {'columns': 6, 'rows': 'auto'}
        }] %}
      {% endif %}
    {% endfor %}

    {# 2. Battery sensors with non-numeric state (unavailable, unknown, etc.) #}
    {% for state in states.sensor
         | rejectattr('attributes.device_class', 'undefined')
         | selectattr('attributes.device_class', '==', 'battery')
         if state.entity_id.endswith('_battery_plus')
            and state.entity_id not in all_exclude_sensors
            and not (state.state | is_number)
            and state.entity_id not in ns.seen %}
      {% set ns.seen = ns.seen + [state.entity_id] %}
      {% set ns.cards = ns.cards + [{
        'type': 'custom:bubble-card',
        'card_type': 'button',
        'button_type': 'state',
        'entity': state.entity_id,
        'grid_options': {'columns': 6, 'rows': 'auto'}
      }] %}
    {% endfor %}

    {# 3. Low binary sensors unavailable/unknown when the matching sensor was
    not already added #} {% for state in states.binary_sensor
         | rejectattr('attributes.device_class', 'undefined')
         | selectattr('attributes.device_class', '==', 'battery')
         if state.entity_id.endswith('_battery_plus_low')
            and state.state in ['unavailable', 'unknown']
            and state.entity_id not in all_exclude_sensors %}
      {% set sensor_id = 'sensor.' ~ state.object_id.removesuffix('_low') %}
      {% set entity = sensor_id if states[sensor_id] is defined else state.entity_id %}
      {% if entity not in all_exclude_sensors and entity not in ns.seen %}
        {% set ns.seen = ns.seen + [entity] %}
        {% set ns.cards = ns.cards + [{
          'type': 'custom:bubble-card',
          'card_type': 'button',
          'button_type': 'state',
          'entity': entity,
          'grid_options': {'columns': 6, 'rows': 'auto'}
        }] %}
      {% endif %}
    {% endfor %}

    {{ ns.cards }}
show_empty: false
```

### Error messages from the browser console

```console

```

### By checking each box below I indicate that I ...

- [x] Understand that this is a channel for reporting bugs, not a support forum (https://community.home-assistant.io/).
- [x] Have made sure I am using the latest version of the plugin.
- [x] Have checked the FAQ section in the README.
- [x] Understand that failure to follow the template above may increase the time required to handle my bug-report, or cause it to be closed without further action.

### PR DESCRIPTION
**Pull request by eyal0**
This fixes #164

### CODE REVIEW

**Comment by eyal0** [comment]
By returning 0 from any comparison against `undefined`, the sort order will look at the next sort config to see what to do.  With this, we can get any order we want:

undefined first and then largest to smallest:

      sort:
        - method: state
          numeric: true
          reverse: true
        - method: state
          reverse: true

largest to smallest and undefined last:

      sort:
        - method: state
          numeric: true
          reverse: true
        - method: state
          reverse: false

smallest to largest and undefined last:

      sort:
        - method: state
          numeric: true
          reverse: false
        - method: state
          reverse: false

undefined and then smallest to largest:

      sort:
        - method: state
          numeric: true
          reverse: false
        - method: state
          reverse: true

It works because first we do the numeric ordering, which gets all the numbers in the order we want but returns `0` for all comparisons against `undefined`, such as "unknown" or "unavailable".  Then, only if we are not comparing two numbers, it falls through to alphabetic.  Numbers come before letters in sort order so without reverse in that second sort config, the numbers come first.  And with reverse in that second sort config, they come last.

**Comment by dcapslock** [comment]
@copilot please review

**Comment by dcapslock** [comment]
I agree with the goal of allowing non-numeric values to fall through to the next sort criterion, but I think Copilot is correct that returning 0 when only one value is non-numeric makes the comparator inconsistent.

How about making placement of non-numeric values an explicit option of numeric sorting:

```yaml 
sort:
  - method: state
    numeric:
      nan: last
  - method: state
``` 

while retaining numeric: true for backwards compatibility.

The comparator could then do something like:

```JavaScript 
const aNan = _a === undefined;
const bNan = _b === undefined;
if (aNan && bNan) return 0;
if (aNan) return method.numeric.nan === "first" ? lt : gt;
if (bNan) return method.numeric.nan === "first" ? gt : lt;
// normal numeric comparison
``` 

This gives non-numeric values a deterministic position relative to numbers, while two non-numeric values compare equal and can therefore be resolved by the next sort criterion.

I prefer nan as an option within numeric rather than a top-level non_numeric option, because it makes clear that this behaviour only applies when performing a numeric sort.

Since nan belongs to the numeric comparison, first/last would determine its final placement; reverse would reverse the numeric values, not change the requested NaN placement.

This expanding of numeric sort config does give rise to editor design challenge but it is worth it to give an understandable config rather than pollution of the main sort config. This is what I was eluding to in declining your first PR. 

Also before any acceptance you will need to include a visual test.

File: src/sort.ts [hunk 1]
```diff
@@ -18,9 +18,7 @@ function compare(_a: any, _b: any, method: SortConfig) {
     }
   }
 
-  if (_a === undefined && _b === undefined) return 0;
-  if (_a === undefined) return gt;
-  if (_b === undefined) return lt;
+  if (_a === undefined || _b === undefined) return 0;
```
[the remark this answers is not in the document]
Comment (eyal0) [reply src/sort.ts]: That's a good point!  The compare function returns -1, 0 or 1.  -1 means before, 1 means after.  In a normal comparison function, the function only returns 0 when items are equal.  What we'd like is a way to indicate that we don't know the relation and we must move to the next.

A reasonable option could be to throw when the comparison is invalid.  So `comp(3,3)==0`, `comp(1,3)<0`, and `comp(undefined, 3)` throw and error because there is no valid result for that.

For the caller: If the function throws, then call the next `SortConfig`.  But if there is no next `SortConfig` then the sorting has failed!  Pick and appropriate failure mechanism in the dashboard.  I don't know if there is a good way to do that.

Worse yet, this will be a change in behavior from before, where undefined always came last.  We don't want to break all those users.  The only situation where behavior has changed to return 0 instead of lt/gt is when one element is undefined and the other isn't.  So, on the last step, for that case, return the previous behavior of lt/gt.

**Comment by dcapslock** [comment]
Hi, I favour the approach I have put forward rather than running by exception and relying on knowledge of list levels in sorting. Exceptions should be for what is not expected. We are expecting non numeric values, they are not exceptions. 

As I understand this issue well I am happy to take up from here to solve your issue.

**Comment by eyal0** [comment]
I pushed another commit that shows how to work around that sort problem.  You have have the comparison throw an exception when the comparison is not valid, like undefined compared with defined.  The caller then deals with the exception.  But that will not work as things used to, which will be a pain.

I'll try your idea.  I want to see how "first" "last" interacts with reverse.  I think that you mean for the "first" "last" to be the same whether or not reverse and I'm not sure if the code as you have written it will work.  Maybe instead of:

```
if (aNan) return method.numeric.nan === "first" ? lt : gt;
if (bNan) return method.numeric.nan === "first" ? gt : lt;
```

You'll need
```
if (aNan) return method.numeric.nan === "first" ? -1 : 1;
if (bNan) return method.numeric.nan === "first" ? 1 : -1;
```

I'll just try it both ways and see what works.

**Comment by dcapslock** [comment]
Ultimately I expect the numeric sort to be self contained with reference to numeric ordering. So to adhere to whatever `nan ` is set. Perhaps an extra step is that the Nans themselves are sorted non numeric within the numeric sort using the same reverse config.

**Comment by eyal0** [comment]
I think that having the key `nan` underneath the numeric key cannot be done in the UI, right?  It would be a shame to break UI configuration just to make it nested...

**Comment by eyal0** [comment]
Okay, I tried the "nan" thing and it doesn't work well in UI.  You can't configure it through the UI properly.

You could change the Numeric to be a choice of three: `off`, `on, nan first`, `on, nan last`.  At least then you can keep the UI.  I'll try that.

**Comment by dcapslock** [comment]
Good progress. Please update the config to be `numeric: nan_first` and `numeric: nan_last` otherwise a yaml config read of `numeric: first` `numeric: last` is confusing.

**Comment by eyal0** [comment]
I tried a few iterations to see how it goes and learn more. Here's what I understand:

Previously, `numeric` could be true or false. False meant to sort alphabetically and true means numeric where all non-numeric items are treated as equal to one another and they are all greater than non numeric.

We'd like a new behavior where the user can decide where non numeric will end up in the case of a numeric search. There are two ways to let the user do this:

1. We could offer the user the option that non numeric is greater/lesser. And the non numerics are all equivalent to one another and all greater or lesser than the numerics.
2. We could offer the user that non numeric is first/last. This is the approach that you mention. It means that the numeric values are the top or the bottom of the list, whether or not that means lesser or greater due to reverse being true/false.

The upside to number 1 is that it is straightforward to implement and consistent with the current code. We would just say that, if numeric is set to "true", that is equivalent to setting it to "nan_greater". The downside to this approach is that it is perhaps more confusing for the user to think in terms of greater/lesser versus first/last.

The upside for approach 2 is that it is straightforward. First/last will be first or last in the list. The downside is that the current behavior in which true is "nan_greater" doesn't map to first nor last. So what do we do in this case?

One option is to just force "numeric: true" to mean last, for example, and some users will experience a change in their display. Or force it to be first and it'll be the opposite set of users that notice the change.

Another way to deal with approach 2 is to continue to allow true to behave as if "nan_greater". So now numeric has false, true, nan_first, and nan_last. (The true behaves like nan_greater.) The UI starts to get unwieldy with options of "off", "on, nan greater", "on, nan first", "on, nan last". I think that it's like the worst of all worlds because it's so confusing.

In my opinion, I'd just offer "false", "nan_greater", and "nan_lesser", approach 1. And the users that currently have true, that'll map to nan_greater. No one will experience any difference. The slight difficulty in understanding that nan_greater acts like first when reverse is true is something that users can experiment with themselves. And the code change is easy without many decisions. Straightforward.

But if you insist on first/last, it can be made to work, you just have to figure out whether you want the complicated config that maintains the previous ordering, using false/true/first/last, or if you want to just break the expectations of some of your users. And if you decide to break expectations of some users, which ones? You have to decide if you want true to mean first or last.

With this issue and all the ramifications understood, how do you want the code to work? I've already coded up every single one so it's just a matter of choosing which version of the code.

**Comment by eyal0** [comment]
> Nans themselves are sorted non numeric within the numeric sort using the same reverse config.

I think better to have all the nans as equivalent to one another so that a follow-up sort could then sort them. Then the user gets full control of where the non numerics go in relation to the numeric and also in what order those two groups are, by layering the SortConfigs.

**Comment by dcapslock** [comment]
As this is getting quite complicated I have suggested an alternative approach for what you are trying to achieve in the issue you posted. 

When I have clear time I will think of alternative approaches as the core of the issue at hand is unavailable of unknown entities. 

One thought is to allow a sorter function specified in JavaScript, though as stated in the issue I suggest you update your template generating the output sorted in the way you wish.

**Comment by dcapslock** [comment]
@eyal0 I have had time to review this closer and request the following changes:

1. change the option yaml config to `nan_first` and `nan_last`
2. to handle legacy, name `true` as "non-numeric follows sort direction". I will include doc notes on this when released.
3. we need visual regression tests but I am happy to do these once you have sorted 1 and 2

Thank you for the submission and patience during reveiw.

**Comment by eyal0** [comment]
Okay, please take another look at the code now.  I don't know much about the visual tests so if you'd like me to undo those and let you do them, let me know and I can back out the last commit.

**Comment by dcapslock** [comment]
Thanks. Tests are perfect. The generated snapshots make part of the repo so I have generated and pushed.

I will get a copilot review - it picks up edge case issues and sleepers, though I don't expect any.

File: src/editor/schema.ts [hunk 2]
```diff
@@ -323,7 +346,17 @@ export const sortSchema = (method) => {
...
-        { name: "numeric", type: "boolean", label: "Numeric sort" },
+        {
+          name: "numeric",
+          type: "select",
+          label: "Numeric sort",
+          options: [
+            [false, "Off"],
+            [true, "On, non-numeric follows sort direction"],
+            ["nan_last", "On, non-numeric last"],
+            ["nan_first", "On, non-numeric first"],
```
[the remark this answers is not in the document]
Comment (dcapslock) [reply src/editor/schema.ts:372]: Addressed in https://github.com/Lint-Free-Technology/lovelace-auto-entities/pull/165/commits/72ed01e4ea9bdbb9c1402c7367f41aaa0c0d92d1

**Comment by eyal0** [comment]
> Preserve the false/off value when serializing the form.

Oh boy, so numeric is supposed to have five options? True, false, nan_first, nan_last, and undefined!

🤣😭🤣😭

**Comment by dcapslock** [comment]
@eyal0 I am looking at this now.

**Comment by eyal0** [comment]
> @eyal0 I am looking at this now.

That seems like it would have been a previous issue when it was just true/false, too.

**Comment by dcapslock** [comment]
> That seems like it would have been a previous issue when it was just true/false, too.

Quite likely. It is a pain but copilot reviews pick up what can be sleeper bugs that turn into support issues.

File: src/editor/schema.ts [hunk 3]
```diff
@@ -292,6 +292,36 @@ export const templateSchema = [
...
+/**
+ * The card configuration distinguishes an omitted numeric setting from an
+ * explicit `false`: timestamp sorts default to numeric only when it is
+ * omitted. Keep that distinction while the GUI form is being edited.
+ */
+export function sortDataForForm(sort?: SortConfig): FormSortConfig {
+  if (!sort) return {};
+  const numeric = sort.numeric === undefined ? "default" : sort.numeric;
+  return { ...sort, numeric };
+}
```
[the remark this answers is not in the document]
Comment (dcapslock) [reply src/editor/schema.ts:316]: Will address

[the remark this answers is not in the document]
Comment (dcapslock) [reply src/editor/schema.ts:316]: Fixed in https://github.com/Lint-Free-Technology/lovelace-auto-entities/pull/165/commits/2c6a0b334acd252d9761ee005c6a9e5c8186d636
